File: src/cz/vutbr/fit/interlockSim/util/HashMapGraph.java
/* Brno University of Technology
 * Faculty of Information Technology
 * 
 * BSc Thesis	2006/2007
 * 
 * Railway Interlocking Simulator
 * 
 * Bedrich Hovorka
 */
package cz.vutbr.fit.interlockSim.util;

import java.util.AbstractCollection;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import java.util.Map.Entry;

/**
 * The ADT ExtendedUnorientedGraph prototype
 *
 * @param <N> nodes
 * @param <E> edges
 * @param <X> {@link ExtendedUnorientedGraph}
 */
public class HashMapGraph<N,E,X> extends AbstractUnorientedGraph<N, E> implements ExtendedUnorientedGraph<N,E,X> {
	final class NodeCollection extends AbstractCollection<N> {
		private final class NodeCollectionIterator implements Iterator<N> {
			private final Iterator<Doubleton<N,X>> keySetIterator = keySet.iterator();
			private Iterator<N> currentPair;
			private N current;

			/**
			 * Construct iterator for iterating over nodes
			 */
			public NodeCollectionIterator() {
				super();
				if (keySetIterator.hasNext()) {
					currentPair = keySetIterator.next().iterator();
				}
			}

			public void remove() {
				if (current == null) throw new IllegalStateException();
				HashMapGraph.this.removeAll(current);
				current = null;
			}

			public N next() {
				if (!currentPair.hasNext()) {
					currentPair = keySetIterator.next().iterator();
				}
				current = currentPair.next();
				return current;
			}

			public boolean hasNext() {
				return (currentPair != null && currentPair.hasNext()) || 
						(keySetIterator != null && keySetIterator.hasNext());
			}
		}

		private final Set<Doubleton<N,X>> keySet = map.keySet();

		@Override
		public Iterator<N> iterator() {
			return new NodeCollectionIterator();
		}

		@Override
		public int size() {
			return keySet.size() << 1;
		}
	}
	
	private NodeCollection nodeCollection;
	
	private HashMap<Doubleton<N,X>, E> map = new HashMap<Doubleton<N,X>, E>();
	
	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#put(N, N, I)
	 */
	public void put(N first, N second, E value) {
	    map.put(new Doubleton<N,X>(first, second), value);
	}
	
	public void put(N first, X firstAddInf, N second, X secondAddInf, E value) {
		map.put(new Doubleton<N,X>(first, second, firstAddInf, secondAddInf), value);
	}
	
	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#get(N, N)
	 */
	public E get(N first, N second) {
	    return map.get(getReferencer(first, second));
	}
	
	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#remove(N, N)
	 */
	public E remove(N first, N second) {
	    return map.remove(getReferencer(first, second));
	}
	
	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#removeAll(N)
	 */
	public Collection<E> removeAll(N node) {
		return allEdgesJoinsWith(node, true);
	}
	
	private Collection<E> allEdgesJoinsWith(N node, boolean remove) {
		assert node != null;
		final Collection<E> collection = new ArrayList<E>();
		
		final Iterator<Entry<Doubleton<N,X>, E>> iterator = map.entrySet().iterator();
		while (iterator.hasNext()) {
			final Entry<Doubleton<N,X>, E> next = iterator.next();
			final Doubleton<N,X> key = next.getKey();
			if (key.contains(node)) {
				collection.add(next.getValue());
				if (remove) iterator.remove();
		    }
		}
		return collection;
	}
	
	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#remove(I)
	 */
	public Collection<N> remove(E h) {
	    assert h != null;
	    final Collection<N> collection = new HashSet<N>();
	    final Iterator<Entry<Doubleton<N,X>, E>> iterator = map.entrySet().iterator();
	    while (iterator.hasNext()) {
	    	final Entry<Doubleton<N,X>, E> next = iterator.next();
	    	if (h.equals(next.getValue())) {
	    		collection.addAll(next.getKey());
	    		iterator.remove();
	    	}
	    }
	    return collection;
	}
	
	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#nodeSet()
	 */
	public Set<N> nodeSet() {
	    if (nodeCollection == null) nodeCollection = new NodeCollection();
		//EXTENSION pohled ne copii
		Set<N> set = new HashSet<N>(nodeCollection);
		//for (Doubleton<N> c: map.keySet()) {
		//	set.addAll(c);
		//}
		return set;
	}

	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#entrySet()
	 */
	public Set<Entry<Doubleton<N,X>, E>> entrySet() {
		return map.entrySet();
	}

	public void putIfNotExists(N first, X addInfFirst, N second, X addInfSecond, E edge) {
		final Doubleton<N,X> pair = new Doubleton<N,X>(first, second, addInfFirst, addInfSecond);  
		if (!map.containsKey(pair)) {
			map.put(pair, edge);
		}
	}
	
	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#values()
	 */
	public Collection<E> values() {
		return map.values();
	}

	public Collection<E> get(N node) {
		return allEdgesJoinsWith(node, false);
	}

	private Doubleton<N, X> getReferencer(N first, N second) {
		return new Doubleton<N, X>(first, second);
	}
	
	public Map<X, E> assignedEdges(final N node) {
		final HashMap<X, E> lmap = new HashMap<X, E>();
		(new DoubletonEntrySetProcessor<X>(map) {
			@Override
			public void processEntryNode(Doubleton<N, X> key, E edge, N node2) {
				if (node.equals(node2)) {
					final X aInf = key.getValue(node);
					assert !lmap.containsKey(aInf);
					lmap.put(aInf, edge); 
				}
			}
		}).process();
		return lmap;
	}

	public X extensionalObject(final N node, final E edge) {
		final DoubletonEntrySetProcessor<X> p = (new DoubletonEntrySetProcessor<X>(map) {
					@Override
					public void processEntryNode(Doubleton<N, X> key, E edge2, N node2) {
						if (node.equals(node2) && edge == edge2) {
							assert getResult() == null;
							setResult(key.getValue(node)); 
						}
					}
				});
		p.process();
		return p.getResult();
	}
	
	abstract class DoubletonEntrySetProcessor <T> {
		private final Map<Doubleton<N, X>, E> map2;

		DoubletonEntrySetProcessor(Map<Doubleton<N, X>, E> map) {
			map2 = map;
		}
		
		private T result;
		
		void process() {
			for (Map.Entry<Doubleton<N, X>, E> e : map2.entrySet()) {
				final Doubleton<N, X> key = e.getKey();
				final Iterator<N> iterator = key.iterator();
				while (iterator.hasNext()) {
					final N next = iterator.next();
					processEntryNode(key, e.getValue(), next);
				}
			}
		}
		
		abstract void processEntryNode(Doubleton<N, X> key, E edge, N node);

		T getResult() {
			return result;
		}

		void setResult(T result) {
			this.result = result;
		}
	}

	public boolean contains(N node1, N node2) {
		return map.containsKey(getReferencer(node1, node2));
	}

	@Override
	protected HashMap<Doubleton<N, X>, E> implementationContainer() {
		return map;
	}
	
	@Override
	public int size() {
		return map.size();
	}
}

File: src/main/kotlin/cz/vutbr/fit/interlockSim/util/HashMapGraph.kt
/* Brno University of Technology
 * Faculty of Information Technology
 *
 * BSc Thesis  2006/2007
 *
 * Railway Interlocking Simulator
 *
 * Bedrich Hovorka
 */
package cz.vutbr.fit.interlockSim.util

import cz.vutbr.fit.interlockSim.exceptions.requireValidState
import java.util.AbstractCollection
import java.util.AbstractSet
import java.util.Collection
import java.util.Iterator
import java.util.Map
import java.util.Map.Entry
import java.util.Set

/**
 * The ADT ExtendedUnorientedGraph prototype
 *
 * @param <N> nodes
 * @param <E> edges
 * @param <X> {@link ExtendedUnorientedGraph}
 */
class HashMapGraph<N, E, X> :
	AbstractUnorientedGraph<N, E>(),
	ExtendedUnorientedGraph<N, E, X> {
	inner class NodeCollection : AbstractCollection<N>() {
		private inner class NodeCollectionIterator : MutableIterator<N> {
			// Safe: Kotlin MutableSet.iterator() returns MutableIterator which is compatible with Java Iterator
			@Suppress("UNCHECKED_CAST")
			private val keySetIterator: java.util.Iterator<Doubleton<N, X>> =
				keySet.iterator() as java.util.Iterator<Doubleton<N, X>>
			private var currentPair: Iterator<N>? = null
			private var current: N? = null

			/**
			 * Construct iterator for iterating over nodes
			 */
			init {
				if (keySetIterator.hasNext()) {
					// Safe: Doubleton.iterator() returns MutableIterator which is compatible with Java Iterator
					@Suppress("UNCHECKED_CAST")
					currentPair = keySetIterator.next().iterator() as Iterator<N>
				}
			}

			override fun remove() {
				if (current == null) throw IllegalStateException()
				this@HashMapGraph.removeAll(current!!)
				current = null
			}

			override fun next(): N {
				if (currentPair?.hasNext() != true) {
					// Safe: Doubleton.iterator() returns MutableIterator which is compatible with Java Iterator
					@Suppress("UNCHECKED_CAST")
					currentPair = keySetIterator.next().iterator() as Iterator<N>
				}
				current = currentPair!!.next()
				return current!!
			}

			override fun hasNext(): Boolean =
				(currentPair != null && currentPair!!.hasNext()) ||
					keySetIterator.hasNext()
		}

		// Safe: Kotlin MutableMap.keys returns MutableSet which is compatible with Java Set
		@Suppress("UNCHECKED_CAST")
		private val keySet: java.util.Set<Doubleton<N, X>> = map.keys as java.util.Set<Doubleton<N, X>>

		override fun iterator(): MutableIterator<N> = NodeCollectionIterator()

		override val size: Int
			get() = keySet.size shl 1
	}

	private var nodeCollection: NodeCollection? = null
	private var cachedNodeSet: java.util.Set<N>? = null

	private val map: MutableMap<Doubleton<N, X>, E> = mutableMapOf()

	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#put(N, N, I)
	 */
	override fun put(
		first: N,
		second: N,
		value: E
	) {
		map[Doubleton<N, X>(first, second)] = value
		// Invalidate cached node set when graph structure changes
		cachedNodeSet = null
	}

	override fun put(
		first: N,
		firstAddInf: X,
		second: N,
		secondAddInf: X,
		value: E
	) {
		map[Doubleton(first, second, firstAddInf, secondAddInf)] = value
		// Invalidate cached node set when graph structure changes
		cachedNodeSet = null
	}

	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#get(N, N)
	 */
	override fun get(
		first: N,
		second: N
	): E? = map[getReferencer(first, second)]

	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#remove(N, N)
	 */
	override fun remove(
		first: N,
		second: N
	): E? {
		val result = map.remove(getReferencer(first, second))
		// Invalidate cached node set when graph structure changes
		if (result != null) cachedNodeSet = null
		return result
	}

	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#removeAll(N)
	 */
	override fun removeAll(node: N): Collection<E> = allEdgesJoinsWith(node, true)

	private fun allEdgesJoinsWith(
		node: N,
		remove: Boolean
	): Collection<E> {
		requireValidState(node != null) { "Node cannot be null" }
		val collection = mutableListOf<E>()

		val iterator = map.entries.iterator()
		var hasRemoved = false
		while (iterator.hasNext()) {
			val next = iterator.next()
			val key = next.key
			if (key.contains(node)) {
				collection.add(next.value)
				if (remove) {
					iterator.remove()
					hasRemoved = true
				}
			}
		}
		// Invalidate cached node set when graph structure changes
		if (hasRemoved) cachedNodeSet = null
		
		// Safe: Kotlin MutableList is compatible with Java Collection
		@Suppress("UNCHECKED_CAST")
		return collection as java.util.Collection<E>
	}

	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#remove(I)
	 */
	override fun remove(h: E): Collection<N> {
		requireValidState(h != null) { "Edge cannot be null" }
		val collection = mutableSetOf<N>()
		val iterator = map.entries.iterator()
		var hasRemoved = false
		while (iterator.hasNext()) {
			val next = iterator.next()
			if (h == next.value) {
				collection.addAll(next.key)
				iterator.remove()
				hasRemoved = true
			}
		}
		// Invalidate cached node set when graph structure changes
		if (hasRemoved) cachedNodeSet = null
		
		// Safe: Kotlin MutableSet is compatible with Java Collection
		@Suppress("UNCHECKED_CAST")
		return collection as java.util.Collection<N>
	}

	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#nodeSet()
	 */
	override fun nodeSet(): Set<N> {
		if (nodeCollection == null) nodeCollection = NodeCollection()
		
		// Lazily build and cache the node set (materializes once, then reuses)
		if (cachedNodeSet == null) {
			@Suppress("UNCHECKED_CAST")
			cachedNodeSet = nodeCollection!!.toMutableSet() as java.util.Set<N>
		}
		
		// Return unmodifiable view of cached set (O(1) performance)
		@Suppress("UNCHECKED_CAST")
		return java.util.Collections.unmodifiableSet(cachedNodeSet as MutableSet<N>) as java.util.Set<N>
	}

	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#entrySet()
	 */
	override fun entrySet(): Set<Map.Entry<Doubleton<N, X>, E>> {
		// Return unmodifiable view of map entries (O(1) performance)
		@Suppress("UNCHECKED_CAST")
		return java.util.Collections.unmodifiableSet(
			map.entries
		) as java.util.Set<Map.Entry<Doubleton<N, X>, E>>
	}

	override fun putIfNotExists(
		first: N,
		addInfFirst: X,
		second: N,
		addInfSecond: X,
		edge: E
	) {
		val pair = Doubleton<N, X>(first, second, addInfFirst, addInfSecond)
		if (!map.containsKey(pair)) {
			map[pair] = edge
			// Invalidate cached node set when graph structure changes
			cachedNodeSet = null
		}
	}

	/* (non-Javadoc)
	 * @see cz.vutbr.fit.interlockSim.context.Graph#values()
	 */
	override fun values(): Collection<E> {
		// Return unmodifiable view of values (O(1) performance)
		@Suppress("UNCHECKED_CAST")
		return java.util.Collections.unmodifiableCollection(
			map.values
		) as java.util.Collection<E>
	}

	override fun get(node: N): Collection<E> = allEdgesJoinsWith(node, false)

	private fun getReferencer(
		first: N,
		second: N
	): Doubleton<N, X> = Doubleton(first, second)

	override fun assignedEdges(node: N): Map<X, E> {
		val lmap = mutableMapOf<X, E>()
		// Safe: Kotlin MutableMap is compatible with Java Map
		@Suppress("UNCHECKED_CAST")
		(
			object : DoubletonEntrySetProcessor<X>(map as java.util.Map<Doubleton<N, X>, E>) {
				override fun processEntryNode(
					key: Doubleton<N, X>,
					edge: E,
					node2: N
				) {
					if (node == node2) {
						val aInf = key.getValue(node)
						requireValidState(!lmap.containsKey(aInf)) { "Duplicate additional info key $aInf for node $node" }
						lmap[aInf!!] = edge
					}
				}
			}
		).process()
		// Safe: Kotlin MutableMap is compatible with Java Map
		@Suppress("UNCHECKED_CAST")
		return lmap as java.util.Map<X, E>
	}

	override fun extensionalObject(
		node: N,
		edge: E
	): X {
		// Safe: Kotlin MutableMap is compatible with Java Map
		@Suppress("UNCHECKED_CAST")
		val p =
			object : DoubletonEntrySetProcessor<X>(map as java.util.Map<Doubleton<N, X>, E>) {
				override fun processEntryNode(
					key: Doubleton<N, X>,
					edge2: E,
					node2: N
				) {
					if (node == node2 && edge == edge2) {
						requireValidState(getResult() == null) { "Multiple extensional objects found for node $node and edge $edge" }
						setResult(key.getValue(node)!!)
					}
				}
			}
		p.process()
		return p.getResult()!!
	}

	abstract inner class DoubletonEntrySetProcessor<T>(
		private val map2: java.util.Map<Doubleton<N, X>, E>
	) {
		private var result: T? = null

		fun process() {
			@Suppress("UNCHECKED_CAST")
			val entries = (map2 as java.util.HashMap<Doubleton<N, X>, E>).entries
			for (e in entries) {
				val key = e.key
				val iterator = key.iterator()
				while (iterator.hasNext()) {
					val next = iterator.next()
					processEntryNode(key, e.value, next)
				}
			}
		}

		abstract fun processEntryNode(
			key: Doubleton<N, X>,
			edge: E,
			node: N
		)

		fun getResult(): T? = result

		fun setResult(result: T) {
			this.result = result
		}
	}

	override fun contains(
		node1: N,
		node2: N
	): Boolean = map.containsKey(getReferencer(node1, node2))

	override fun implementationContainer(): HashMap<Doubleton<N, X>, E> = map as java.util.HashMap<Doubleton<N, X>, E>

	override fun size(): Int = map.size

	override fun clear() {
		map.clear()
		// Invalidate cached node set when graph structure changes
		cachedNodeSet = null
	}
}
